### Import Dependancies

In [2]:
import json
import pandas as pd
import numpy as np
import re

In [3]:
import os
os.getcwd()

'C:\\Users\\User\\Documents\\Data Analytics Bootcamp\\Course Modules\\Unit 3 - Databases\\Module 8 - ETL\\DABC-Movies-ETL-V2'

#### Create a variable to hold the directory that holds the data

In [4]:
file_dir ='C:/Users/User/Documents/Data Analytics Bootcamp/Course Modules/Unit 3 - Databases/Module 8 - ETL/DABC-Movies-ETL-v2/Resources'

### Load the JSON file into a List of Dictionaries

In [5]:
with open(f'{file_dir}/wikipedia-movies.json', mode='r') as file:
    wiki_movies_raw = json.load(file)

In [6]:
len(wiki_movies_raw)

7311

#### Inspect the data

In [7]:
# First 5 records
wiki_movies_raw[:5]

[{'url': 'https://en.wikipedia.org/wiki/The_Adventures_of_Ford_Fairlane',
  'year': 1990,
  'imdb_link': 'https://www.imdb.com/title/tt0098987/',
  'title': 'The Adventures of Ford Fairlane',
  'Directed by': 'Renny Harlin',
  'Produced by': ['Steve Perry', 'Joel Silver'],
  'Screenplay by': ['David Arnott', 'James Cappe', 'Daniel Waters'],
  'Story by': ['David Arnott', 'James Cappe'],
  'Based on': ['Characters', 'by Rex Weiner'],
  'Starring': ['Andrew Dice Clay',
   'Wayne Newton',
   'Priscilla Presley',
   'Lauren Holly',
   'Morris Day',
   'Robert Englund',
   "Ed O'Neill"],
  'Narrated by': 'Andrew "Dice" Clay',
  'Music by': ['Cliff Eidelman', 'Yello'],
  'Cinematography': 'Oliver Wood',
  'Edited by': 'Michael Tronick',
  'Productioncompany ': 'Silver Pictures',
  'Distributed by': '20th Century Fox',
  'Release date': ['July 11, 1990', '(', '1990-07-11', ')'],
  'Running time': '102 minutes',
  'Country': 'United States',
  'Language': 'English',
  'Budget': '$20 million',


In [8]:
# Last 5 records
wiki_movies_raw[-5:]

[{'url': 'https://en.wikipedia.org/wiki/Holmes_%26_Watson',
  'year': 2018,
  'imdb_link': 'https://www.imdb.com/title/tt1255919/',
  'title': 'Holmes & Watson',
  'Directed by': 'Etan Cohen',
  'Produced by': ['Will Ferrell',
   'Adam McKay',
   'Jimmy Miller',
   'Clayton Townsend'],
  'Screenplay by': 'Etan Cohen',
  'Based on': ['Sherlock Holmes',
   'and',
   'Dr. Watson',
   'by',
   'Sir Arthur Conan Doyle'],
  'Starring': ['Will Ferrell',
   'John C. Reilly',
   'Rebecca Hall',
   'Rob Brydon',
   'Steve Coogan',
   'Ralph Fiennes'],
  'Music by': 'Mark Mothersbaugh',
  'Cinematography': 'Oliver Wood',
  'Edited by': 'Dean Zimmerman',
  'Productioncompanies ': ['Columbia Pictures',
   'Gary Sanchez Productions',
   'Mosaic Media Group',
   'Mimran Schur Pictures'],
  'Distributed by': 'Sony Pictures Releasing',
  'Release date': ['December 25, 2018',
   '(',
   '2018-12-25',
   ')',
   '(United States)'],
  'Running time': '90 minutes',
  'Country': 'United States',
  'Language

In [9]:
# Some records in the middle
wiki_movies_raw[3600:3605]

[{'url': 'https://en.wikipedia.org/wiki/Benji:_Off_the_Leash!',
  'year': 2004,
  'imdb_link': 'https://www.imdb.com/title/tt0315273/',
  'title': 'Benji: Off the Leash!',
  'Directed by': 'Joe Camp',
  'Written by': 'Joe Camp',
  'Starring': ['Benji', 'Nick Whitaker', 'Shaggy', 'Gypsy the Cockatoo'],
  'Music by': 'Antonio di Lorenzo',
  'Productioncompany ': 'Mulberry Square Productions',
  'Distributed by': 'Mulberry Square Productions',
  'Release date': ['March 26, 2004', '(', '2004-03-26', ')'],
  'Running time': '97 min',
  'Country': 'United States',
  'Language': 'English',
  'Box office': '$3,817,362'},
 {'url': 'https://en.wikipedia.org/wiki/The_Best_Thief_in_the_World',
  'year': 2004,
  'imdb_link': 'https://www.imdb.com/title/tt0389796/',
  'title': 'The Best Thief in the World',
  'Directed by': 'Jacob Kornbluth',
  'Produced by': ['Tim Perrell', 'Nicola Usborne'],
  'Written by': 'Jacob Kornbluth',
  'Starring': ['Marc Rozendaal',
   'Michael Silverman',
   'David Warsh

#### Create a dataframe from raw data and inspect it

In [10]:
wiki_movies_df = pd.DataFrame(wiki_movies_raw)
wiki_movies_df.head()

url    year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990.0   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990.0   
2   https://en.wikipedia.org/wiki/Air_America_(film)  1990.0   
3    https://en.wikipedia.org/wiki/Alice_(1990_film)  1990.0   
4      https://en.wikipedia.org/wiki/Almost_an_Angel  1990.0   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2  https://www.imdb.com/title/tt0099005/                      Air America   
3  https://www.imdb.com/title/tt0099012/                            Alice   
4  https://www.imdb.com/title/tt0099018/                  Almost an Angel   

          Directed by                  Produced by  \
0        Renny Harlin   [Steve Perry, Joel Silver]   
1         James Foley  [Ric Kidney, Robert Redlin]   
2  Roger Spottiswoode               Daniel Melnick   
3         Woody Allen              Robert Greenhut   
4        John Cornell                 John Cornell   

                                Screenplay by                     Story by  \
0  [David Arnott, James Cappe, Daniel Waters]  [David Arnott, James Cappe]   
1                [James Foley, Robert Redlin]                          NaN   
2                  [John Eskow, Richard Rush]                          NaN   
3                                         NaN                          NaN   
4                                         NaN                          NaN   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   
2             [Air America, by, Christopher Robbins]   
3                                                NaN   
4                                                NaN   

                                            Starring  ... Predecessor  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...  ...         NaN   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...  ...         NaN   
2  [Mel Gibson, Robert Downey Jr., Nancy Travis, ...  ...         NaN   
3  [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...  ...         NaN   
4        [Paul Hogan, Elias Koteas, Linda Kozlowski]  ...         NaN   

  Founders Area served Products Services Russian Hebrew Revenue  \
0      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
1      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
2      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
3      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
4      NaN         NaN      NaN      NaN     NaN    NaN     NaN   

  Operating income Polish  
0              NaN    NaN  
1              NaN    NaN  
2              NaN    NaN  
3              NaN    NaN  
4              NaN    NaN  

[5 rows x 193 columns]

##### Convert wiki_movies_df.columns to a list, so we can see all of the columns

In [11]:
wiki_movies_df.columns.tolist()

['url',
 'year',
 'imdb_link',
 'title',
 'Directed by',
 'Produced by',
 'Screenplay by',
 'Story by',
 'Based on',
 'Starring',
 'Narrated by',
 'Music by',
 'Cinematography',
 'Edited by',
 'Productioncompany ',
 'Distributed by',
 'Release date',
 'Running time',
 'Country',
 'Language',
 'Budget',
 'Box office',
 'Written by',
 'Genre',
 'Theme music composer',
 'Country of origin',
 'Original language(s)',
 'Producer(s)',
 'Editor(s)',
 'Production company(s)',
 'Original network',
 'Original release',
 'Productioncompanies ',
 'Executive producer(s)',
 'Production location(s)',
 'Distributor',
 'Picture format',
 'Audio format',
 'Voices of',
 'Followed by',
 'Composer(s)',
 'Created by',
 'Also known as',
 'Opening theme',
 'No. of episodes',
 'Preceded by',
 'Author',
 'Publisher',
 'Publication date',
 'Media type',
 'Pages',
 'ISBN',
 'OCLC',
 'LC Class',
 'Cover artist',
 'Series',
 'Set in',
 'Adaptation by',
 'Suggested by',
 'Biographical data',
 'Born',
 'Died',
 'Resti

### Use List Comprehensions to Filter Data

Create a list comprehension with the filter expression and save that to an intermediate variable wiki_movies

- The filter expression filters for 


1. rows that have values in either the 'Director' or 'Directed by' columns, 


2. have values in the 'imdb_link' column,


3. and have no values for 'No. of episodes' to filter out TV shows.

In [12]:
wiki_movies = [movie for movie in wiki_movies_raw
               if ('Director' in movie or 'Directed by' in movie)
                   and 'imdb_link' in movie
                   and 'No. of episodes' not in movie]
len(wiki_movies)

7076

In [13]:
wiki_movies_filt_dt = pd.DataFrame(wiki_movies)
wiki_movies_filt_dt.head()

url  year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990   
2   https://en.wikipedia.org/wiki/Air_America_(film)  1990   
3    https://en.wikipedia.org/wiki/Alice_(1990_film)  1990   
4      https://en.wikipedia.org/wiki/Almost_an_Angel  1990   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2  https://www.imdb.com/title/tt0099005/                      Air America   
3  https://www.imdb.com/title/tt0099012/                            Alice   
4  https://www.imdb.com/title/tt0099018/                  Almost an Angel   

          Directed by                  Produced by  \
0        Renny Harlin   [Steve Perry, Joel Silver]   
1         James Foley  [Ric Kidney, Robert Redlin]   
2  Roger Spottiswoode               Daniel Melnick   
3         Woody Allen              Robert Greenhut   
4        John Cornell                 John Cornell   

                                Screenplay by                     Story by  \
0  [David Arnott, James Cappe, Daniel Waters]  [David Arnott, James Cappe]   
1                [James Foley, Robert Redlin]                          NaN   
2                  [John Eskow, Richard Rush]                          NaN   
3                                         NaN                          NaN   
4                                         NaN                          NaN   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   
2             [Air America, by, Christopher Robbins]   
3                                                NaN   
4                                                NaN   

                                            Starring  ... Hepburn Literally  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...  ...     NaN       NaN   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...  ...     NaN       NaN   
2  [Mel Gibson, Robert Downey Jr., Nancy Travis, ...  ...     NaN       NaN   
3  [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...  ...     NaN       NaN   
4        [Paul Hogan, Elias Koteas, Linda Kozlowski]  ...     NaN       NaN   

  Cantonese Chinese Yiddish Arabic Romanized Russian Hebrew Polish  
0       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
1       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
2       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
3       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  
4       NaN     NaN     NaN    NaN       NaN     NaN    NaN    NaN  

[5 rows x 75 columns]

#### Next, let's take a look at the different languages.

Filter for movies where there is value for the column "Arabic"

In [14]:
wiki_movies_df[wiki_movies_df['Arabic'].notnull()]

url    year  \
7060  https://en.wikipedia.org/wiki/The_Insult_(film)  2018.0   
7293   https://en.wikipedia.org/wiki/Capernaum_(film)  2018.0   

                                  imdb_link       title    Directed by  \
7060  https://www.imdb.com/title/tt7048622/  The Insult   Ziad Doueiri   
7293  https://www.imdb.com/title/tt8267604/   Capernaum  Nadine Labaki   

                                            Produced by  \
7060  [Rachid Bouchareb, Jean Bréhat, Julie Gayet, A...   
7293                    [Michel Merkt, Khaled Mouzanar]   

                                          Screenplay by  \
7060                                                NaN   
7293  [Nadine Labaki, Jihad Hojaily, Michelle Keserw...   

                                               Story by Based on  \
7060                                                NaN      NaN   
7293  [Georges Khabbaz, Nadine Labaki, Michelle Kese...      NaN   

                                               Starring  ... Predecessor  \
7060                       [Adel Karam, Kamel El Basha]  ...         NaN   
7293  [Zain Al Rafeea, Yordanos Shiferaw, Boluwatife...  ...         NaN   

     Founders Area served Products Services Russian Hebrew Revenue  \
7060      NaN         NaN      NaN      NaN     NaN    NaN     NaN   
7293      NaN         NaN      NaN      NaN     NaN    NaN     NaN   

     Operating income Polish  
7060              NaN    NaN  
7293              NaN    NaN  

[2 rows x 193 columns]

Two movies are returned. Let's grab the movie's Wikipedia page (url) for details.

In [15]:
wiki_movies_df[wiki_movies_df['Arabic'].notnull()]['url']

7060    https://en.wikipedia.org/wiki/The_Insult_(film)
7293     https://en.wikipedia.org/wiki/Capernaum_(film)
Name: url, dtype: object

The language columns contain alternate titles in different languages.

Sort the column names by displaying the columns in a list, and then inspect which columns contain alternate titles in different languages.

The below sorted() function will sort it in alphabetical order.

In [16]:
sorted(wiki_movies_df.columns.tolist())

['Actor control',
 'Adaptation by',
 'Alias',
 'Alma mater',
 'Also known as',
 'Animation by',
 'Arabic',
 'Area',
 'Area served',
 'Artist(s)',
 'Attraction type',
 'Audio format',
 'Author',
 'Based on',
 'Biographical data',
 'Bopomofo',
 'Born',
 'Box office',
 'Budget',
 'Camera setup',
 'Cantonese',
 'Characters',
 'Children',
 'Chinese',
 'Cinematography',
 'Closing date',
 'Color process',
 'Comics',
 'Composer(s)',
 'Coordinates',
 'Country',
 'Country of origin',
 'Cover artist',
 'Created by',
 'Date premiered',
 'Designer(s)',
 'Developed by',
 'Developer(s)',
 'Dewey Decimal',
 'Died',
 'Directed by',
 'Director',
 'Distributed by',
 'Distributor',
 'Divisions',
 'Duration',
 'Edited by',
 'Editor(s)',
 'Ending theme',
 'Engine',
 'Engine(s)',
 'Executive producer(s)',
 'Family',
 'Fate',
 'Film(s)',
 'Followed by',
 'Format(s)',
 'Formerly',
 'Founded',
 'Founder',
 'Founders',
 'French',
 'Full name',
 'Gender',
 'Genre',
 'Genre(s)',
 'Genres',
 'Gwoyeu Romatzyh',
 'Ha

In [17]:
wiki_movies_df[wiki_movies_df['Language'].notnull()]

url    year  \
0     https://en.wikipedia.org/wiki/The_Adventures_o...  1990.0   
1     https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990.0   
2      https://en.wikipedia.org/wiki/Air_America_(film)  1990.0   
3       https://en.wikipedia.org/wiki/Alice_(1990_film)  1990.0   
4         https://en.wikipedia.org/wiki/Almost_an_Angel  1990.0   
...                                                 ...     ...   
7306    https://en.wikipedia.org/wiki/Holmes_%26_Watson  2018.0   
7307     https://en.wikipedia.org/wiki/Vice_(2018_film)  2018.0   
7308  https://en.wikipedia.org/wiki/On_the_Basis_of_Sex  2018.0   
7309  https://en.wikipedia.org/wiki/Destroyer_(2018_...  2018.0   
7310  https://en.wikipedia.org/wiki/Black_Mirror:_Ba...  2018.0   

                                  imdb_link                            title  \
0     https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1     https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2     https://www.imdb.com/title/tt0099005/                      Air America   
3     https://www.imdb.com/title/tt0099012/                            Alice   
4     https://www.imdb.com/title/tt0099018/                  Almost an Angel   
...                                     ...                              ...   
7306  https://www.imdb.com/title/tt1255919/                  Holmes & Watson   
7307  https://www.imdb.com/title/tt6266538/                             Vice   
7308  https://www.imdb.com/title/tt4669788/              On the Basis of Sex   
7309  https://www.imdb.com/title/tt7137380/                        Destroyer   
7310  https://www.imdb.com/title/tt9495224/                     Bandersnatch   

             Directed by                                        Produced by  \
0           Renny Harlin                         [Steve Perry, Joel Silver]   
1            James Foley                        [Ric Kidney, Robert Redlin]   
2     Roger Spottiswoode                                     Daniel Melnick   
3            Woody Allen                                    Robert Greenhut   
4           John Cornell                                       John Cornell   
...                  ...                                                ...   
7306          Etan Cohen  [Will Ferrell, Adam McKay, Jimmy Miller, Clayt...   
7307          Adam McKay  [Brad Pitt, Dede Gardner, Jeremy Kleiner, Kevi...   
7308          Mimi Leder                                     Robert W. Cort   
7309        Karyn Kusama             [Fred Berger, Phil Hay, Matt Manfredi]   
7310         David Slade                                     Russell McLean   

                                   Screenplay by                     Story by  \
0     [David Arnott, James Cappe, Daniel Waters]  [David Arnott, James Cappe]   
1                   [James Foley, Robert Redlin]                          NaN   
2                     [John Eskow, Richard Rush]                          NaN   
3                                            NaN                          NaN   
4                                            NaN                          NaN   
...                                          ...                          ...   
7306                                  Etan Cohen                          NaN   
7307                                         NaN                          NaN   
7308                                         NaN                          NaN   
7309                                         NaN                          NaN   
7310                                         NaN                          NaN   

                                               Based on  \
0                           [Characters, by Rex Weiner]   
1     [the novel, After Dark, My Sweet, by, Jim Thom...   
2                [Air America, by, Christopher Robbins]   
3                                                   NaN   
4                                                   NaN   
...       

Here is a complete list of columns that hold alternate title data: 

Also known as, Arabic, Cantonese, Chinese, French, Hangul, Hebrew, Hepburn, Japanese, Literally,  Mandarin, McCune–Reischauer, Original title, Polish,  Revised Romanization, Romanized, Russian, Simplified, Traditional, Yiddish

##### Create a non-destructive copy of the list of movies

Inside of the function, we can create a new local variable called "movie" and assign it the new copy of the parameter "movie".

This way, inside of the clean_movie() function, "movie" will refer to the local copy. 
Any changes we make inside clean_movie() will now only affect the copy, so if we make a mistake, we still have the original, untouched "movie" to reference.

In [18]:
def clean_movie(movie):
    
    # create a non-destructive copy
    movie_copy = dict(movie)
    
    # create an empty dict to hold alternative titles
    alt_titles = {}
    
    
    # Loop through a list of all alternative title keys
    for key in ['Also known as','Arabic','Cantonese','Chinese','French',
                'Hangul','Hebrew','Hepburn','Japanese','Literally',
                'Mandarin','McCune–Reischauer','Original title','Polish',
                'Revised Romanization','Romanized','Russian',
                'Simplified','Traditional','Yiddish']:
        
        # Check if the current key exists in the movie object
        if key in movie:
            
            # If so, remove the key-value pair and add to the alternative titles dictionary.
            alt_titles[key] = movie[key]
            movie.pop(key)
            
    # After looping through every key, add the alternative titles dict to the movie object.        
    if len(alt_titles) > 0:
        movie['alt_titles'] = alt_titles    
    
    # Merge column names
    def change_column_name(old_name, new_name):
        if old_name in movie:
            movie[new_name]=movie.pop(old_name)
    change_column_name('Adaptation by', 'Writer(s)')
    change_column_name('Country of origin', 'Country')
    change_column_name('Directed by', 'Director')
    change_column_name('Distributed by', 'Distributor')
    change_column_name('Edited by', 'Editor(s)')
    change_column_name('Length', 'Running time')
    change_column_name('Original release', 'Release date')
    change_column_name('Music by', 'Composer(s)')
    change_column_name('Produced by', 'Producer(s)')
    change_column_name('Producer', 'Producer(s)')
    change_column_name('Productioncompanies ', 'Production company(s)')
    change_column_name('Productioncompany ', 'Production company(s)')
    change_column_name('Released', 'Release Date')
    change_column_name('Release Date', 'Release date')
    change_column_name('Screen story by', 'Writer(s)')
    change_column_name('Screenplay by', 'Writer(s)')
    change_column_name('Story by', 'Writer(s)')
    change_column_name('Theme music composer', 'Composer(s)')
    change_column_name('Written by', 'Writer(s)')    
    
    return movie

Make a list of cleaned movies with a list comprehension:

In [19]:
clean_movies = [clean_movie(movie) for movie in wiki_movies]

Set wiki_movies_df to be the DataFrame created from clean_movies, and print out a list of the columns.

In [20]:
wiki_movies_df = pd.DataFrame(clean_movies)
sorted(wiki_movies_df.columns.tolist())

['Animation by',
 'Audio format',
 'Based on',
 'Box office',
 'Budget',
 'Cinematography',
 'Color process',
 'Composer(s)',
 'Country',
 'Created by',
 'Director',
 'Distributor',
 'Editor(s)',
 'Executive producer(s)',
 'Followed by',
 'Genre',
 'Label',
 'Language',
 'Narrated by',
 'Original language(s)',
 'Original network',
 'Picture format',
 'Preceded by',
 'Producer(s)',
 'Production company(s)',
 'Production location(s)',
 'Recorded',
 'Release date',
 'Running time',
 'Starring',
 'Suggested by',
 'Venue',
 'Voices of',
 'Writer(s)',
 'alt_titles',
 'imdb_link',
 'title',
 'url',
 'year']

#### Remove Duplicated Rows

In order to merge the DF with the data from Kaggle, we need to ensure that there are no duplicated rows, as identified by IMDb ID.

- To extract the ID, we need to use Regular Expressions (regex).

Let's use regular expressions in Pandas' built-in string methods that work on a Series object accessed with the str property.

- str.extract(), which takes in a regular expression pattern

IMDb links generally look like "https://www.imdb.com/title/tt1234567/", with "tt1234567" as the IMDb ID.
 
- The regular expression for a group of characters that start with "tt" and has seven digits is "(tt\d{7})"

- Since regular expressions use backslashes, which Python also uses for special characters, we want to tell Python to treat our regular expression characters as a raw string of text. Therefore, we put an r before the quotes. 

In [21]:
wiki_movies_df['imdb_id'] = wiki_movies_df['imdb_link'].str.extract(r'(tt\d{7})')
print(len(wiki_movies_df))

# Use the drop_duplicates() method to remove duplicates, specify subset as imdb_id since that should be unique
wiki_movies_df.drop_duplicates(subset='imdb_id', inplace=True)
print(len(wiki_movies_df))

wiki_movies_df.head()

7076
7033


url  year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990   
2   https://en.wikipedia.org/wiki/Air_America_(film)  1990   
3    https://en.wikipedia.org/wiki/Alice_(1990_film)  1990   
4      https://en.wikipedia.org/wiki/Almost_an_Angel  1990   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2  https://www.imdb.com/title/tt0099005/                      Air America   
3  https://www.imdb.com/title/tt0099012/                            Alice   
4  https://www.imdb.com/title/tt0099018/                  Almost an Angel   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   
2             [Air America, by, Christopher Robbins]   
3                                                NaN   
4                                                NaN   

                                            Starring         Narrated by  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...  Andrew "Dice" Clay   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...                 NaN   
2  [Mel Gibson, Robert Downey Jr., Nancy Travis, ...                 NaN   
3  [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...                 NaN   
4        [Paul Hogan, Elias Koteas, Linda Kozlowski]                 NaN   

   Cinematography                                       Release date  \
0     Oliver Wood                  [July 11, 1990, (, 1990-07-11, )]   
1    Mark Plummer  [May 17, 1990, (, 1990-05-17, ), (Cannes Film ...   
2   Roger Deakins                [August 10, 1990, (, 1990-08-10, )]   
3  Carlo Di Palma              [December 25, 1990, (, 1990-12-25, )]   
4    Russell Boyd                                  December 19, 1990   

  Running time  ... Created by Preceded by Suggested by alt_titles Recorded  \
0  102 minutes  ...        NaN         NaN          NaN        NaN      NaN   
1  114 minutes  ...        NaN         NaN          NaN        NaN      NaN   
2  113 minutes  ...        NaN         NaN          NaN        NaN      NaN   
3  106 minutes  ...        NaN         NaN          NaN        NaN      NaN   
4   95 minutes  ...        NaN         NaN          NaN        NaN      NaN   

  Venue Label Animation by Color process    imdb_id  
0   NaN   NaN          NaN           NaN  tt0098987  
1   NaN   NaN          NaN           NaN  tt0098994  
2   NaN   NaN          NaN           NaN  tt0099005  
3   NaN   NaN          NaN           NaN  tt0099012  
4   NaN   NaN          NaN           NaN  tt0099018  

[5 rows x 40 columns]

### Remove Mostly Null Columns

One way to get the count of null values for each column is to use a list comprehension

In [22]:
[[column,wiki_movies_df[column].isnull().sum()] for column in wiki_movies_df.columns]

[['url', 0],
 ['year', 0],
 ['imdb_link', 0],
 ['title', 1],
 ['Based on', 4852],
 ['Starring', 184],
 ['Narrated by', 6752],
 ['Cinematography', 691],
 ['Release date', 32],
 ['Running time', 139],
 ['Country', 236],
 ['Language', 244],
 ['Budget', 2295],
 ['Box office', 1548],
 ['Director', 0],
 ['Distributor', 357],
 ['Editor(s)', 548],
 ['Composer(s)', 518],
 ['Producer(s)', 202],
 ['Production company(s)', 1678],
 ['Writer(s)', 199],
 ['Genre', 6923],
 ['Original language(s)', 6875],
 ['Original network', 6908],
 ['Executive producer(s)', 6936],
 ['Production location(s)', 6986],
 ['Picture format', 6969],
 ['Audio format', 6972],
 ['Voices of', 7031],
 ['Followed by', 7024],
 ['Created by', 7023],
 ['Preceded by', 7023],
 ['Suggested by', 7032],
 ['alt_titles', 7012],
 ['Recorded', 7031],
 ['Venue', 7032],
 ['Label', 7031],
 ['Animation by', 7031],
 ['Color process', 7032],
 ['imdb_id', 0]]

 Let's make a list of columns that have less than 90% null values and use those to trim down our dataset, which gives us the columns that we want to keep.

In [23]:
len(wiki_movies_df)

7033

In [24]:
[column for column in wiki_movies_df.columns if wiki_movies_df[column].isnull().sum() < len(wiki_movies_df) * 0.9]

['url',
 'year',
 'imdb_link',
 'title',
 'Based on',
 'Starring',
 'Cinematography',
 'Release date',
 'Running time',
 'Country',
 'Language',
 'Budget',
 'Box office',
 'Director',
 'Distributor',
 'Editor(s)',
 'Composer(s)',
 'Producer(s)',
 'Production company(s)',
 'Writer(s)',
 'imdb_id']

In [25]:
wiki_columns_to_keep = [column for column in wiki_movies_df.columns if wiki_movies_df[column].isnull().sum() < len(wiki_movies_df) * 0.9]
wiki_movies_df = wiki_movies_df[wiki_columns_to_keep]

In [26]:
wiki_movies_df.head()

url  year  \
0  https://en.wikipedia.org/wiki/The_Adventures_o...  1990   
1  https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990   
2   https://en.wikipedia.org/wiki/Air_America_(film)  1990   
3    https://en.wikipedia.org/wiki/Alice_(1990_film)  1990   
4      https://en.wikipedia.org/wiki/Almost_an_Angel  1990   

                               imdb_link                            title  \
0  https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1  https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2  https://www.imdb.com/title/tt0099005/                      Air America   
3  https://www.imdb.com/title/tt0099012/                            Alice   
4  https://www.imdb.com/title/tt0099018/                  Almost an Angel   

                                            Based on  \
0                        [Characters, by Rex Weiner]   
1  [the novel, After Dark, My Sweet, by, Jim Thom...   
2             [Air America, by, Christopher Robbins]   
3                                                NaN   
4                                                NaN   

                                            Starring  Cinematography  \
0  [Andrew Dice Clay, Wayne Newton, Priscilla Pre...     Oliver Wood   
1  [Jason Patric, Rachel Ward, Bruce Dern, George...    Mark Plummer   
2  [Mel Gibson, Robert Downey Jr., Nancy Travis, ...   Roger Deakins   
3  [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...  Carlo Di Palma   
4        [Paul Hogan, Elias Koteas, Linda Kozlowski]    Russell Boyd   

                                        Release date Running time  \
0                  [July 11, 1990, (, 1990-07-11, )]  102 minutes   
1  [May 17, 1990, (, 1990-05-17, ), (Cannes Film ...  114 minutes   
2                [August 10, 1990, (, 1990-08-10, )]  113 minutes   
3              [December 25, 1990, (, 1990-12-25, )]  106 minutes   
4                                  December 19, 1990   95 minutes   

         Country  ...       Budget        Box office            Director  \
0  United States  ...  $20 million     $21.4 million        Renny Harlin   
1  United States  ...   $6 million      $2.7 million         James Foley   
2  United States  ...  $35 million       $57,718,089  Roger Spottiswoode   
3  United States  ...  $12 million        $7,331,647         Woody Allen   
4             US  ...  $25 million  $6,939,946 (USA)        John Cornell   

          Distributor                       Editor(s)  \
0    20th Century Fox                 Michael Tronick   
1     Avenue Pictures                 Howard E. Smith   
2    TriStar Pictures  [John Bloom, Lois Freeman-Fox]   
3      Orion Pictures                  Susan E. Morse   
4  Paramount Pictures                    David Stiven   

               Composer(s)                  Producer(s)  \
0  [Cliff Eidelman, Yello]   [Steve Perry, Joel Silver]   
1            Maurice Jarre  [Ric Kidney, Robert Redlin]   
2            Charles Gross               Daniel Melnick   
3                      NaN              Robert Greenhut   
4            Maurice Jarre                 John Cornell   

                   Production company(s)                     Writer(s)  \
0                        Silver Pictures   [David Arnott, James Cappe]   
1                        Avenue Pictures  [James Foley, Robert Redlin]   
2  [Carolco Pictures, IndieProd Company]    [John Eskow, Richard Rush]   
3                                    NaN                   Woody Allen   
4                                    NaN                    Paul Hogan   

     imdb_id  
0  tt0098987  
1  tt0098994  
2  tt0099005  
3  tt0099012  
4  tt0099018  

[5 rows x 21 columns]

### Check datatype for each column, convert  and parse the data

In [27]:
wiki_movies_df.dtypes

url                      object
year                      int64
imdb_link                object
title                    object
Based on                 object
Starring                 object
Cinematography           object
Release date             object
Running time             object
Country                  object
Language                 object
Budget                   object
Box office               object
Director                 object
Distributor              object
Editor(s)                object
Composer(s)              object
Producer(s)              object
Production company(s)    object
Writer(s)                object
imdb_id                  object
dtype: object

Below columns will need to be adjusted:
- Box office should be numeric.
- Budget should be numeric.
- Release date should be a date object.
- Running time should be numeric.

#### Parsing Box Office Data

It will be helpful to only look at rows where box office data is defined, so first we'll make a data series that drops missing values with the following:

In [28]:
box_office = wiki_movies_df['Box office'].dropna()

In [29]:
len(box_office)

5485

Regular expressions only work on strings, so we'll need to make sure all of the box office data is entered as a string.

By using the apply() method, we can see which values are not strings.

#### Method A

First, make a is_not_a_string() function:

    def is_not_a_string(x):
        return type(x) != str
    
Then add the following:    
    
    box_office[box_office.map(is_not_a_string)]

Method B

Use a lambda function instead of defining a function in a name

    lambda x: type(x) != str)
    
Pass in this lamda function into the map() function

In [30]:
box_office[box_office.map(lambda x: type(x) != str)]

34                           [US$, 4,212,828]
54      [$6,698,361 (, United States, ), [2]]
74                    [$6,488,144, (US), [1]]
126                [US$1,531,489, (domestic)]
130                          [US$, 4,803,039]
                        ...                  
6980               [$99.6, million, [4], [5]]
6994                   [$365.6, million, [1]]
6995                         [$53.8, million]
7015                     [$435, million, [7]]
7048                   [$529.3, million, [4]]
Name: Box office, Length: 135, dtype: object

Methods A and B would return the same results; 
Method B avoids the need to define a new function everytime we want to use the map() method.

There are quite a few data points that are stored as lists.

We will use the join() string method to concat list items into one string.

- Note: join() method will need a separator string. We will use a space (" ") as the separator

combine the apply() method with the lambda function to join list items into one string


In [31]:
box_office = box_office.apply(lambda x: ' '.join(x) if type(x) == list else x)

In [32]:
box_office.head()

0       $21.4 million
1        $2.7 million
2         $57,718,089
3          $7,331,647
4    $6,939,946 (USA)
Name: Box office, dtype: object

As shown above, there are many styles present.

We'll use regular expressions to find out just how many of each style are in our data.

To do so, we will use the built-in Python module for regular expressions: re

import the library on the first cell, and re-run.

    import re

There are two main forms the box office data is written in: "$123.4 million" (or billion), and "$123,456,789." 

We're going to build a regular expression for each form, and then see what forms are left over.

For the first form, our pattern match string will include six elements in the following order:

1. A dollar sign
2. An arbitrary (but non-zero) number of digits
3. An optional decimal point
4. An arbitrary (but possibly zero) number of more digits
5. A space (maybe more than one)
6. The word "million" or "billion"

Check out this YouTube video for REGEX tutorial.

https://www.youtube.com/watch?v=K8L6KVGG-7o&ab_channel=CoreySchafer

In [33]:
# Note: this replacement is added here to address the issues arising from the first attempt, 
# Issue addressed :  Some values use a period as a thousands separator, not a comma.
box_office = box_office.str.replace(r'\$.*[-—–](?![a-z])', '$', regex=True)

Create a variable form_one and set it equal to the finished regular expression string.

- Because we need the escape characters to remain, we need to preface the string with an r.

In [34]:
# Note: this is the revised version of form_one after the first attempt to address the issues illustrated in sections below.
form_one = r'\$\s*\d+\.?\d*\s*[mb]illi?on'

Now, to count up how many box office values match our first form.

- We'll use the str.contains() method on box_office.

- To ignore whether letters are uppercase or lowercase, add an argument called flags, and set it equal to re.IGNORECASE.

- Finally, we can call the sum()method to count up the total number that return True.

In [35]:
box_office.str.contains(form_one, flags=re.IGNORECASE).sum()

3909

In the first attempt*

- There are 3,896 box office values that match the form "$123.4 million/billion." 

In the second attempt with the revised version:

- There are 3,909 box office values that match the form "$123.4 million/billion." and addresses the spaces in between the dollar sign and the number, and miss spelling.

Next, we'll match the numbers of our second form, "$123,456,789." 

In words, our pattern match string will include the following elements:

1. A dollar sign
2. A group of one to three digits
3. least one group starting with a comma and followed by exactly three digits

Create another variable form_two and set it equal to the finished regular expression string.

Don't forget to make it a raw string so Python keeps the escaped characters.

In [36]:
# Note: this is the revised version of form_two after the first attempt to address the issues illustrated in sections below.

form_two = r'\$\s*\d{1,3}(?:[,\.]\d{3})+(?!\s[mb]illion)'
box_office.str.contains(form_two, flags=re.IGNORECASE).sum()

1559

In the first attempt:

- There are 1,544 box office values that match the form "$123,456,789."

In the second attempt:

- There are 1,559 matches and addressed the issues illustrated in sections below.

#### Most of the box office values are described by either form. Now we want to see which values aren't described by either.

In [37]:
# create two Boolean Series
matches_form_one = box_office.str.contains(form_one, flags=re.IGNORECASE)
matches_form_two = box_office.str.contains(form_two, flags=re.IGNORECASE)

Pandas has element-wise logical operators:

- The element-wise negation operator is the tilde: ~ (similar to "not")
- The element-wise logical "and" is the ampersand: &
- The element-wise logical "or" is the pipe: |

In [38]:
# This is the revised output after making changes to the form_one and form_two variables to fix the below 4 issues.

box_office[~matches_form_one & ~matches_form_two]

600                      $5000 (US)
1070                     35,254,617
1480                     £3 million
1865                   ¥1.1 billion
2032                            N/A
2091                           $309
2665    926,423 admissions (France)
3631                            TBA
3879       CN¥3.650 million (China)
4116                     £7,385,434
4306                            $30
4561        $45.2k (only in Turkey)
5447                          £2.56
5784                       413 733$
6013                        Unknown
6369                          $111k
6370                           $588
6593                 less than $372
6843                        8 crore
Name: Box office, dtype: object

We can fix our pattern matches to capture more values by addressing these issues:

1. Some values have spaces in between the dollar sign and the number.
2. Some values use a period as a thousands separator, not a comma.
3. Some values are given as a range.
4. "Million" is sometimes misspelled as "millon."

For references, our previous expresision for form one and form two are as follows:

    form_one = r'\$\d+\.?\d*\s*[mb]illion'
    form_two = r'\$\d{1,3}(?:,\d{3})+'
    
To address the above 4 issues, We will change the expressions to:

    form_one = r'\$\s*\d+\.?\d*\s*[mb]illi?on'
    form_two = r'\$\s*\d{1,3}(?:[,\.]\d{3})+(?!\s[mb]illion)'
    
    and we will add the below replacement before defining form_one
    
    box_office = box_office.str.replace(r'\$.*[-—–](?![a-z])', '$', regex=True)

After making the changes, the list of exceptions is much shorter. These are too small to account for; leave them.

### Extract and Convert the Box Office Values

Now that we've got expressions to match almost all the box office values, we'll use them to extract only the parts of the strings that match. 

The f-string 

    f'{form_one}|{form_two}' 
    
will create a regular expression that matches either form_one or form_two, so we just need to put the whole thing in parentheses to create a capture group

In [39]:
box_office.str.extract(f'({form_one}|{form_two})')

0
0     $21.4 million
1      $2.7 million
2       $57,718,089
3        $7,331,647
4        $6,939,946
...             ...
7070  $19.4 million
7071  $41.9 million
7072  $76.1 million
7073  $38.4 million
7074   $5.5 million

[5485 rows x 1 columns]

Now we need a function to turn the extracted values into a numeric value.

We'll call it parse_dollars, and parse_dollars will take in a string and return a floating-point number. 

    def parse_dollars(s):
        # if s is not a string, return NaN

        # if input is of the form $###.# million

            # remove dollar sign and " million"

            # convert to float and multiply by a million

            # return value

        # if input is of the form $###.# billion

            # remove dollar sign and " billion"

            # convert to float and multiply by a billion

            # return value

        # if input is of the form $###,###,###

            # remove dollar sign and commas

            # convert to float

            # return value

        # otherwise, return NaN

Since we're working directly with strings, we'll use the re module to access the regular expression functions.

We'll use re.match(pattern, string) to see if our string matches a pattern

In [40]:
def parse_dollars(s):
    # if s is not a string, return NaN
    if type(s) != str:
        return np.nan

    # if input is of the form $###.# million
    if re.match(r'\$\s*\d+\.?\d*\s*milli?on', s, flags=re.IGNORECASE):

        # remove dollar sign and " million"
        s = re.sub('\$|\s|[a-zA-Z]','', s)

        # convert to float and multiply by a million
        value = float(s) * 10**6

        # return value
        return value

    # if input is of the form $###.# billion
    elif re.match(r'\$\s*\d+\.?\d*\s*billi?on', s, flags=re.IGNORECASE):

        # remove dollar sign and " billion"
        s = re.sub('\$|\s|[a-zA-Z]','', s)

        # convert to float and multiply by a billion
        value = float(s) * 10**9

        # return value
        return value

    # if input is of the form $###,###,###
    elif re.match(r'\$\s*\d{1,3}(?:[,\.]\d{3})+(?!\s[mb]illion)', s, flags=re.IGNORECASE):

        # remove dollar sign and commas
        s = re.sub('\$|,','', s)

        # convert to float
        value = float(s)

        # return value
        return value

    # otherwise, return NaN
    else:
        return np.nan

Now we have everything we need to parse the box office values to numeric values.

- First, we need to extract the values from box_office using str.extract.

- Then we'll apply parse_dollars to the first column in the DataFrame returned by str.extract

Below is the final code from these 2 parts:

In [41]:
wiki_movies_df['box_office'] = box_office.str.extract(f'({form_one}|{form_two})', flags=re.IGNORECASE)[0].apply(parse_dollars)

In [42]:
wiki_movies_df['box_office']

0       21400000.0
1        2700000.0
2       57718089.0
3        7331647.0
4        6939946.0
           ...    
7071    41900000.0
7072    76100000.0
7073    38400000.0
7074     5500000.0
7075           NaN
Name: box_office, Length: 7033, dtype: float64

We no longer need the Box Office column, so we'll just drop it:

In [43]:
wiki_movies_df.drop('Box office', axis=1, inplace=True)

#### Parsing Budget Data

Use the same pattern matches and see how many budget values are in a different form.

Create a budget variable:

In [44]:
budget = wiki_movies_df['Budget'].dropna()

Convert any lists to strings:

- again, using lambda functions

In [45]:
budget = budget.map(lambda x: ' '.join(x) if type(x) == list else x)

Then remove any values between a dollar sign and a hyphen (for budgets given in ranges):

In [46]:
budget = budget.str.replace(r'\$.*[-—–](?![a-z])', '$', regex=True)

Use the same pattern matches that you created to parse the box office data, and apply them without modifications to the budget data. Then, look at what's left.

In [47]:
matches_form_one = budget.str.contains(form_one, flags=re.IGNORECASE)
matches_form_two = budget.str.contains(form_two, flags=re.IGNORECASE)
budget[~matches_form_one & ~matches_form_two]

136                         Unknown
204     60 million Norwegian Kroner
478                         Unknown
973             $34 [3] [4] million
1126               $120 [4] million
1226                        Unknown
1278                            HBO
1374                     £6,000,000
1397                     13 million
1480                   £2.8 million
1734                   CAD2,000,000
1913     PHP 85 million (estimated)
1948                    102,888,900
1953                   3,500,000 DM
1973                     ₤2,300,874
2281                     $14 milion
2451                     ₤6,350,000
3144                   € 40 million
3360               $150 [6] million
3418                        $218.32
3802                   £4.2 million
3906                            N/A
3959                    760,000 USD
4470                       19 crore
4641                    £17 million
5034              $$200 [4] million
5055           $155 [2] [3] million
5419                $40 [4] 

That parsed almost all of the budget data. However, there's a new issue with the budget data: citation references (the numbers in square brackets).

Remove the citation references with a regular expression:

In [48]:
budget = budget.str.replace(r'\[\d+\]\s*', '')
budget[~matches_form_one & ~matches_form_two]

136                         Unknown
204     60 million Norwegian Kroner
478                         Unknown
973                     $34 million
1126                   $120 million
1226                        Unknown
1278                            HBO
1374                     £6,000,000
1397                     13 million
1480                   £2.8 million
1734                   CAD2,000,000
1913     PHP 85 million (estimated)
1948                    102,888,900
1953                   3,500,000 DM
1973                     ₤2,300,874
2281                     $14 milion
2451                     ₤6,350,000
3144                   € 40 million
3360                   $150 million
3418                        $218.32
3802                   £4.2 million
3906                            N/A
3959                    760,000 USD
4470                       19 crore
4641                    £17 million
5034                  $$200 million
5055                   $155 million
5419                    $40 

This is ready for parsing. The remaining items are not worth the time to fix.

In [49]:
wiki_movies_df['budget'] = budget.str.extract(f'({form_one}|{form_two})', flags=re.IGNORECASE)[0].apply(parse_dollars)

Drop the original Budget column.

In [50]:
wiki_movies_df.drop('Budget', axis=1, inplace=True)

#### Parse Release Date

First, make a variable that holds the non-null values of Release date in the DataFrame, converting lists to strings:

In [51]:
release_date = wiki_movies_df['Release date'].dropna().apply(lambda x: ' '.join(x) if type(x) == list else x)

The forms we'll be parsing are:

1. Full month name, one- to two-digit day, four-digit year (i.e., January 1, 2000)


2. Four-digit year, two-digit month, two-digit day, with any separator (i.e., 2000-01-01)


3. Full month name, four-digit year (i.e., January 2000)


4. Four-digit year


In [52]:
# matches the month, dd, yyyy format
date_form_one = r'(?:January|February|March|April|May|June|July|August|September|October|November|December)\s[123]\d,\s\d{4}'

# matches yyyy.mm.dd
date_form_two = r'\d{4}.[01]\d.[123]\d'

# matches month yyyy
date_form_three = r'(?:January|February|March|April|May|June|July|August|September|October|November|December)\s\d{4}'

# matches yyyy
date_form_four = r'\d{4}'

For longer expressions, consider using verbose REGEX to comment on each component on the expressions.

https://stackoverflow.com/questions/13851794/how-to-implement-a-verbose-regex-in-python

Extract the dates 

In [53]:
release_date.str.extract(f'({date_form_one}|{date_form_two}|{date_form_three}|{date_form_four})', flags=re.IGNORECASE)

0
0         July 11, 1990
1          May 17, 1990
2       August 10, 1990
3     December 25, 1990
4     December 19, 1990
...                 ...
7071  December 25, 2018
7072  December 11, 2018
7073               2018
7074    August 31, 2018
7075      December 2018

[7001 rows x 1 columns]

Instead of creating our own function to parse the dates, we'll use the built-in

    to_datetime() 
    
method in Pandas. 

Since there are different date formats, set the 

    infer_datetime_format

option to True. 


The date formats we've targeted are among those that the to_datetime() function can recognize, which explains the infer_datetime_format=True argument below.

In [55]:
wiki_movies_df['release_date'] = pd.to_datetime(release_date.str.extract(f'({date_form_one}|{date_form_two}|{date_form_three}|{date_form_four})')[0], infer_datetime_format=True)

#### Parse Running Time

First, make a variable that holds the non-null values of Release date in the DataFrame, converting lists to strings:

In [57]:
running_time = wiki_movies_df['Running time'].dropna().apply(lambda x: ' '.join(x) if type(x) == list else x)

In [58]:
running_time

0                                 102 minutes
1                                 114 minutes
2                                 113 minutes
3                                 106 minutes
4                                  95 minutes
                        ...                  
7071                               90 minutes
7072                              132 minutes
7073                              120 minutes
7074                              123 minutes
7075    Variable; 90 minutes for default path
Name: Running time, Length: 6894, dtype: object

It looks like most of the entries just look like "100 minutes." 

Let's see how many running times look exactly like that by using string boundaries

In [60]:
running_time.str.contains(r'^\d*\s*minutes$', flags=re.IGNORECASE).sum()

6528

Let's get a sense of what the other 366 entries look like. 

(Total Length of series 6,894 - above # of records 6,528 = 366)

In [61]:
running_time[running_time.str.contains(r'^\d*\s*minutes$', flags=re.IGNORECASE) != True]

9                                                 102 min
26                                                 93 min
28                                                32 min.
34                                                101 min
35                                                 97 min
                              ...                        
6500       114 minutes [1] 120 minutes (extended edition)
6643                                             104 mins
6709    90 minutes (theatrical) [1] 91 minutes (unrate...
7057    108 minutes (Original cut) 98 minutes (UK cut)...
7075                Variable; 90 minutes for default path
Name: Running time, Length: 366, dtype: object

Let's make this more general by only marking the beginning of the string, and accepting other abbreviations of "minutes" by only searching up to the letter "m."

In [62]:
running_time.str.contains(r'^\d*\s*m', flags=re.IGNORECASE).sum()

6877

That accounts for 6,877 entries. The remaining 17 follow:

In [63]:
running_time[running_time.str.contains(r'^\d*\s*m', flags=re.IGNORECASE) != True]

668                     UK:84 min (DVD version) US:86 min
727                         78-102 min (depending on cut)
840                       Varies (79 [3] –84 [1] minutes)
1347                                              25 : 03
1443    United States: 77 minutes Argentina: 94 minute...
1499                                            1hr 35min
1551                                               varies
1774                    Netherlands:96 min, Canada:95 min
1777                                       approx. 14 min
2273                                           1 h 43 min
2993                                               1h 48m
3925                                              4 hours
4425    US domestic version: 86 minutes Original versi...
4967    Theatrical cut: 97 minutes Unrated cut: 107 mi...
5424                    115 [1] /123 [2] /128 [3] minutes
5447                                    1 hour 32 minutes
7075                Variable; 90 minutes for default path
Name: Running 

We can capture some more of these by relaxing the condition that the pattern has to start with at the beginning of the string, but the entries with hours and minutes listed separately will give erroneous data.

In [64]:
running_time.str.contains(r'\d*\s*m', flags=re.IGNORECASE).sum()

6891

We can match all of the hour + minute patterns with one regular expression pattern. Our pattern follows:

1. Start with one or more digits.


2. Have an optional space after the digit and before the letter "h."


3. Capture all the possible abbreviations of "hour(s)." To do this, we'll make every letter in "hours" optional except the "h."


4. Have an optional space after the "hours" marker.


5. Have an optional number of digits for minutes.


As a pattern, this looks like

    "\d+\s*ho?u?r?s?\s*\d*"

In [65]:
running_time.str.contains(r'\d+\s*ho?u?r?s?\s*\d*', flags=re.IGNORECASE).sum()

5

Extract values to allow for both possible patterns.

- add capture groups around the \d instances
- add an alternating character

In [66]:
running_time_extract = running_time.str.extract(r'(\d+)\s*ho?u?r?s?\s*(\d*)|(\d+)\s*m')

Unfortunately, this new DataFrame is all strings, we'll need to convert them to numeric values.

Because we may have captured empty strings, we'll use the 

    to_numeric()
    
method and set the errors argument to 'coerce'.
    
- Coercing the errors will turn the empty strings into Not a Number (NaN), then we can use fillna() to change all the NaNs to zeros.

In [69]:
running_time_extract = running_time_extract.apply(lambda col: pd.to_numeric(col, errors='coerce')).fillna(0)

Now we can apply a function that will convert the hour capture groups and minute capture groups to minutes if the pure minutes capture group is zero, and save the output to wiki_movies_df:

In [70]:
wiki_movies_df['running_time'] = running_time_extract.apply(lambda row: row[0]*60 + row[1] if row[2] == 0 else row[2], axis=1)

Finally, we can drop Running time from the dataset with the following code:

In [71]:
wiki_movies_df.drop('Running time', axis=1, inplace=True)

In [72]:
wiki_movies_df

url  year  \
0     https://en.wikipedia.org/wiki/The_Adventures_o...  1990   
1     https://en.wikipedia.org/wiki/After_Dark,_My_S...  1990   
2      https://en.wikipedia.org/wiki/Air_America_(film)  1990   
3       https://en.wikipedia.org/wiki/Alice_(1990_film)  1990   
4         https://en.wikipedia.org/wiki/Almost_an_Angel  1990   
...                                                 ...   ...   
7071    https://en.wikipedia.org/wiki/Holmes_%26_Watson  2018   
7072     https://en.wikipedia.org/wiki/Vice_(2018_film)  2018   
7073  https://en.wikipedia.org/wiki/On_the_Basis_of_Sex  2018   
7074  https://en.wikipedia.org/wiki/Destroyer_(2018_...  2018   
7075  https://en.wikipedia.org/wiki/Black_Mirror:_Ba...  2018   

                                  imdb_link                            title  \
0     https://www.imdb.com/title/tt0098987/  The Adventures of Ford Fairlane   
1     https://www.imdb.com/title/tt0098994/             After Dark, My Sweet   
2     https://www.imdb.com/title/tt0099005/                      Air America   
3     https://www.imdb.com/title/tt0099012/                            Alice   
4     https://www.imdb.com/title/tt0099018/                  Almost an Angel   
...                                     ...                              ...   
7071  https://www.imdb.com/title/tt1255919/                  Holmes & Watson   
7072  https://www.imdb.com/title/tt6266538/                             Vice   
7073  https://www.imdb.com/title/tt4669788/              On the Basis of Sex   
7074  https://www.imdb.com/title/tt7137380/                        Destroyer   
7075  https://www.imdb.com/title/tt9495224/                     Bandersnatch   

                                               Based on  \
0                           [Characters, by Rex Weiner]   
1     [the novel, After Dark, My Sweet, by, Jim Thom...   
2                [Air America, by, Christopher Robbins]   
3                                                   NaN   
4                                                   NaN   
...                                                 ...   
7071  [Sherlock Holmes, and, Dr. Watson, by, Sir Art...   
7072                                                NaN   
7073                                                NaN   
7074                                                NaN   
7075                                                NaN   

                                               Starring  \
0     [Andrew Dice Clay, Wayne Newton, Priscilla Pre...   
1     [Jason Patric, Rachel Ward, Bruce Dern, George...   
2     [Mel Gibson, Robert Downey Jr., Nancy Travis, ...   
3     [Alec Baldwin, Blythe Danner, Judy Davis, Mia ...   
4           [Paul Hogan, Elias Koteas, Linda Kozlowski]   
...                                                 ...   
7071  [Will Ferrell, John C. Reilly, Rebecca Hall, R...   
7072  [Christian Bale, Amy Adams, Steve Carell, Sam ...   
7073  [Felicity Jones, Armie Hammer, Justin Theroux,...   
7074  [Nicole Kidman, Sebastian Stan, Toby Kebbell, ...   
7075  [Fionn Whitehead, Will Poulter, Craig Parkinso...   

                     Cinematography  \
0                       Oliver Wood   
1                      Mark Plummer   
2                     Roger Deakins   
3                    Carlo Di Palma   
4                      Russell Boyd   
...                             ...   
7071                    Oliver Wood   
7072                   Greig Fraser   
7073                  Michael Grady   
7074                 Julie Kirkwood   
7075  [Aaron Morton, Jake Polonsky]   

                                           Release date         Country  \
0                     [July 11, 1990, (, 1990-07-11, )]   United States   
1     [May 17, 1990, (, 1990-05-17, ), (Cannes Film ...   United States   
2                   [August 10, 1990, (, 1990-08-10, )]   United States   
3                 [December 25, 1990, (, 1990-12-25, )]   United States   
4                                     December 1

In [74]:
wiki_movies_df.to_csv('CleanedData/wiki_extract.csv')